CPP/DesktopToastsCppApp/DesktopNotificationManagerCompat.cpp

```diff
@@ -0,0 +1,288 @@
+const wchar_t *DesktopNotificationManagerCompat::s_aumid = nullptr;
+bool DesktopNotificationManagerCompat::s_registeredActivator = false;
+bool DesktopNotificationManagerCompat::s_hasCheckedIsRunningAsUwp = false;
+bool DesktopNotificationManagerCompat::s_isRunningAsUwp = false;
+
+HRESULT DesktopNotificationManagerCompat::RegisterAumidAndComServer(const wchar_t *aumid, GUID clsid)
+{
+    // If running as Desktop Bridge
+    if (IsRunningAsUwp())
+    {
+        // Clear the AUMID since Desktop Bridge doesn't use it, and then we're done.
+        // Desktop Bridge apps are registered with platform through their manifest.
+        // Their LocalServer32 key is also registered through their manifest.
+        s_aumid = nullptr;
+        s_registeredAumidAndComServer = true;
+        return S_OK;
+    }
+
+    s_aumid = aumid;
+
+    // Get the EXE path
+    wchar_t exePath[MAX_PATH];
+    DWORD charWritten = ::GetModuleFileName(nullptr, exePath, ARRAYSIZE(exePath));
+    auto hr = charWritten > 0 ? S_OK : HRESULT_FROM_WIN32(::GetLastError());
+    if (SUCCEEDED(hr))
+    {
+        // Register the COM server
+        hr = RegisterComServer(clsid, exePath);
+        if (SUCCEEDED(hr))
+        {
+            s_registeredAumidAndComServer = true;
+        }
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::RegisterActivator()
+{
+    // Module<OutOfProc> needs a callback registered before it can be used.
+    // Since we don't care about when it shuts down, we'll pass an empty lambda here.
+    Module<OutOfProc>::Create([] {});
+
+    // If a local server process only hosts the COM object then COM expects
+    // the COM server host to shutdown when the references drop to zero.
+    // Since the user might still be using the program after activating the notification,
+    // we don't want to shutdown immediately.  Incrementing the object count tells COM that
+    // we aren't done yet.
+    Module<OutOfProc>::GetModule().IncrementObjectCount();
+
+    auto hr = Module<OutOfProc>::GetModule().RegisterObjects();
+    if (SUCCEEDED(hr))
+    {
+        s_registeredActivator = true;
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::RegisterComServer(GUID clsid, wchar_t exePath[])
+{
+    // Turn the GUID into a string
+    OLECHAR* clsidOlechar;
+    StringFromCLSID(clsid, &clsidOlechar);
+    std::wstring clsidStr(clsidOlechar);
+    ::CoTaskMemFree(clsidOlechar);
+
+    // Create the subkey
+    // Something like SOFTWARE\Classes\CLSID\{23A5B06E-20BB-4E7E-A0AC-6982ED6A6041}\LocalServer32
+    std::wstring subKeyStr = LR"(SOFTWARE\Classes\CLSID\)" + clsidStr + LR"(\LocalServer32)";
+    LPCWSTR subKey = subKeyStr.c_str();
+
+    // Include -ToastActivated launch args on the exe
+    std::wstring exePathStr(exePath);
+    exePathStr = L"\"" + exePathStr + L"\" -ToastActivated";
+    exePath = new wchar_t[exePathStr.length() + 1];
```

Comment:
This leaks the exePath that was passed in, and is a very odd pattern.
Is the goal to return this variable? If so make it a separate parameter.

Reply:
The goal is just to...

1. Surround the EXE path in quotes
2. Add the ` -ToastActivated` flag to the end of the string

Using the std string libraries was the easiest way I could figure out that concatenation, what's the right way to do it? :)

Reply:
my comment wasn't actually correct, you're leaking the exePath you allocate on line 84. Either way, you should be able to just use exePathStr.c_str() directly instead of copying it.
As for exePathStr, you are mostly fine, but you should remove the initialization on line 82. line 83 makes you reallocate anyway, no need to do it an extra time.